lib/rubocop/cop/style/hash_transform_methods.rb

```diff
@@ -0,0 +1,222 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for uses of `_.each_with_object({}) {...}`,
+      # `_.map{...}.to_h`, and `Hash[_.map{...}]` that are actually just
+      # transforming either the keys or the values of a hash, and tries to use
+      # a simpler & faster call to `transform_keys` or `transform_values`
+      # instead.
+      #
+      # This can produce false positives if we are transforming an enumerable
+      # of key-value-like pairs that isn't actually a hash, e.g.:
+      # `[[k1, v1], [k2, v2], ...]`
+      #
+      # This cop should only be enabled on Ruby version 2.5 or newer
+      # (`transform_values` was added in 2.4, and `transform_keys` in 2.5.)
+      #
+      # @example
+      #   # bad
+      #   {a: 1, b: 2}.each_with_object({}) { |(k, v), h| h[k] = v*v }
+      #   {a: 1, b: 2}.map { |k, v| [k.to_s, v] }
+      #
+      #   # good
+      #   {a: 1, b: 2}.transform_values { |v| v*v }
+      #   {a: 1, b: 2}.transform_keys { |k| k.to_s }
+      class HashTransformMethods < Cop
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        EACH_WITH_OBJECT_PATTERN = NodePattern.new(<<~PATTERN)
+          (block
+            ({send csend} _ :each_with_object (hash))
+            (args
+              (mlhs
+                (arg $_)
+                (arg $_))
+              (arg $_))
+            ({send csend} (lvar $_) :[]= $_ $_))
+        PATTERN
+
+        HASH_BRACKETS_MAP_PATTERN = NodePattern.new(<<~PATTERN)
+          (send
+            (const _ :Hash)
+            :[]
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_)))
+        PATTERN
+
+        MAP_TO_H_PATTERN = NodePattern.new(<<~PATTERN)
+          ({send csend}
+            (block
+              ({send csend} _ :map)
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_))
+            :to_h)
+        PATTERN
+
+        VALS_MSG = 'Use `transform_values` to transform just the values of a ' \
```

Comment:
Don't forget `.`s to terminate those messages. A better wording would be `Prefer ... over ....` No more details are needed, as that's how most messages are structured today.